decoder/decoder.go

```diff
@@ -21,18 +22,21 @@ type DecoderStruct struct {
 	eth     layers.Ethernet
 	ip4     layers.IPv4
 	ip6     layers.IPv6
+	icmp4   layers.ICMPv4
+	icmp6   layers.ICMPv6
 	tcp     layers.TCP
 	udp     layers.UDP
 	payload gopacket.Payload
 	decoded []gopacket.LayerType
 
-	tcpProc tcp.Processor
-	udpProc udp.Processor
+	icmpProc icmp.Processor
```

Comment:
there's quite some overlap between ICMPv4 and v6, but at this point we might consider them different -> icmp4Proc, icmp6Proc. Also split icmp.Processor interface definition.